nimbess-cni.go

```diff
@@ -0,0 +1,314 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Main package for the Nimbess CNI plugin
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"time"
+
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/version"
+	"google.golang.org/grpc"
+
+	cnitypes "github.com/containernetworking/cni/pkg/types/current"
+	bv "github.com/containernetworking/plugins/pkg/utils/buildversion"
+	cninimbess "github.com/nimbess/nimbess-agent/pkg/proto/cni"
+	log "github.com/sirupsen/logrus"
+)
+
+const defaultLogFile = "/tmp/nimbess-cni.log"
+
+// NimbessConfig is whatever you expect your configuration json to be. This is whatever
+// is passed in on stdin. Your plugin may wish to expose its functionality via
+// runtime args, see CONVENTIONS.md in the CNI spec.
+type nimbessConfig struct {
+	// common CNI config
+	types.NetConf
+
+	// PrevResult contains previous plugin's result, used only when called in the context of a chained plugin.
+	PrevResult *map[string]interface{} `json:"prevResult"`
+
+	// GrpcServer is a plugin-specific config, contains location of the gRPC server
+	// where the CNI requests are being forwarded to (server:port tuple, e.g. "localhost:9111")
+	// or unix-domain socket path (e.g. "/run/cni.sock").
+	GrpcServer string `json:"grpcServer"`
+
+	// LogFile is a plugin-specific config, specifies location of the CNI plugin log file.
+	// If empty, plugin logs into defaultLogFile.
+	LogFile string `json:"logFile"`
+
+	// EtcdEndpoints is a plugin-specific config, may contain comma-separated list of ETCD endpoints
+	// required for specific for the CNI / IPAM plugin.
+	EtcdEndpoints string `json:"etcdEndpoints"`
+}
+
+// grpcConnect sets up a connection to the gRPC server specified in grpcServer argument
+// as a server:port tuple (e.g. "localhost:9111") or unix-domain socket path (e.g. "/run/cni.sock").
+func grpcConnect(grpcServer string) (conn *grpc.ClientConn, cni cninimbess.RemoteCNIClient, err error) {
+
+	if grpcServer != "" && grpcServer[0] == '/' {
+		// unix-domain socket connection
+		conn, err = grpc.Dial(
+			grpcServer,
+			grpc.WithInsecure(),
+			grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+				return net.DialTimeout("unix", addr, timeout)
+			}),
+		)
+	} else {
+		// TCP socket connection
+		conn, err = grpc.Dial(grpcServer, grpc.WithInsecure())
+	}
+
+	if err != nil {
+		return
```

Comment:
does this compile - it doesn't return any of the three expected return values.

Reply:
It does compile. As it's pretty common in Go world, exception catching is done by comparing if error is not null. This is not checking if the grpc connection is not empty, or has some format, it is checking if any of the conditions above hasn't throw an error.

Reply:
The return values for the function written have three return values (conn *grpc.ClientConn, cni cninimbess.RemoteCNIClient, err error) , so thought it would not compile if values would not be returned for all three.

Reply:
I see what you mean now! :-) It didn't complain at least.

Reply:
The code is right. If you used named parameters, return will return all of the values without specifying them in the return statement. If you had only specified type parameters it would error. For example, this will error:
https://play.golang.org/p/T1BNnppBJeN

But this will not:
https://play.golang.org/p/pqIOX45It83

Reply:
Ooh another goodie to stash away. I didn't realize naming the return values actually creates local variables for them so you can use them. Very cool.